### ISSUE TITLE
Configuration

### ISSUE THREAD

**Issue by FichteFoll** [knitori/shanghai#8]
Use YAML for configuration, parser: [ruamel.yaml](https://pypi.python.org/pypi/ruamel.yaml) (persists comments and style when modifying in-place).

Latest structure proposal: https://paste.chireiden.net/q/ODbUrfpWNFcT

-

- [ ] general parsing
- [ ] auto-reload on changes
  - [ ] detect which networks have changed by comparing against cached versions and only dis-/re-/connect those
  - [ ] provide an event for plugins they can hook

**Comment by FichteFoll** [comment]
Imo we should follow the IRC URI spec of only prepending hashes for channels that start with an alphanumeric character. [[1]](https://en.wikipedia.org/wiki/Internet_Relay_Chat#URI_scheme)

**Comment by knitori** [comment]
Forgot to reference, but pr #13 adds yaml config parsing.

**Comment by FichteFoll** [comment]
Note to self: Prepend `#` to channels in config only if they start with an alphanumeric character, as described [here](https://en.wikipedia.org/wiki/Internet_Relay_Chat#URI_scheme).

**Comment by knitori** [comment]
I actually [did that](https://github.com/chireiden/shanghai/pull/13/files#diff-381c4cef5168f7297efa4b0460abb16eR75), but the other way around - only prepend `#` if it doesn't start with any of `#`, `&`, `+` and `!`, since a channel can start with non-ascii (e.g. utf-8) chars on some networks like euIRC.

**Comment by FichteFoll** [comment]
Yeah well, it's good enough. I know that ZNC uses a channel named `~#znc` as partyline, which would be impossible to configure this way however. Maybe intercept the server's ` CHANTYPES=#&+~` message?

**Comment by knitori** [comment]
Right, but then the prepending has to be deferred to after receiving the 005 message (and consequently after the 001, where I usually issued the channel join commands).
